### PR TITLE
Homework3 task1 - cache decorator function implementation with tests

### PR DESCRIPTION
**Pull request by IuliiaLanchuk**

### CODE REVIEW

File: homework3/task01/project1/task1.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+def cache(times: int) -> Callable:
+    """
+    Remember other function output value. Give out cached value up to times number.
+
+    :times: quantity of returned cached values
+    """
+
+    def wrapper(func: Callable) -> Any:
+        current_time = 0
```
Comment (Cjkjvfnby) [inline homework3/task01/project1/task1.py]: Please check `current time` translation.

File: homework3/task01/tests/test_cach_decorator.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+from project1.task1 import cache
+
+
+@cache(times=2)
+def f(a: int):
+    return a + 1
+
+
+def test_cache_decorator():
+    assert f(6) is f() is f()
```
Comment (Cjkjvfnby) [inline homework3/task01/tests/test_cach_decorator.py]: It is expected that you should cache the function result for each argument separately. 

```python
@cache(times=1)
def foo(arg):
   print(arg)
   return arg
```
In that case, if you call

foo(1)
foo(1)
foo(2)
foo(1)

You will get 3 lines in the console
1 (first call with 1)
2 (first call with 2)
1 (third call with 1)

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]
I'd recommend finishing with HW 2 task 4 before proceeding with that task.

File: homework3/task01/project_cache/task1cache.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+"""
+Task.
+
+In previous homework 2 task 4, you wrote a cache function that remembers other function output value.
```
Comment (Cjkjvfnby) [inline homework3/task01/project_cache/task1cache.py:4]: I'd recommend waiting `homework 2 task 4` before continuing with this task.

File: homework3/task01/project_cache/task1cache.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    """
+    Remember other function output value. Give out cached value up to times number.
+
+    :times: quantity of returned cached values
+    """
+
+    def wrapper(func: Callable) -> Callable:
+        cache_data = {}
+        if times < 1:
+            raise ValueError("Times value should be more than zero")
```
Comment (Cjkjvfnby) [inline homework3/task01/project_cache/task1cache.py]: `more than zero` usually referenced as `positive`.

File: homework3/task01/project_cache/task1cache.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+        def timer(*args: Any) -> Any:
```
Comment (Cjkjvfnby) [inline homework3/task01/project_cache/task1cache.py]: I'd say this name usually used for different purposes, so maybe name it like `cache_with_counter`

File: homework3/task01/project_cache/task1cache.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            if args in cache_data and cache_data[args][1] < times:
+                cache_data[args][1] += 1
```
Comment (Cjkjvfnby) [inline homework3/task01/project_cache/task1cache.py]: Usually, it is more convenient to set the times at the beginning and decrement till zero.

File: homework3/task01/tests/test_cach_decorator.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+    assert captured.out == "6\n6\n"
+
+
+def test_cache_func_with_mock_after_4th_func_call_cache_was_cleared(capsys):
+    @cache(times=3)
+    def foo(*arg):
+        return arg
+
+    a, b, c, d, f = foo(Mock), foo(Mock), foo(Mock), foo(Mock), foo(Mock)
```
Comment (Cjkjvfnby) [inline homework3/task01/tests/test_cach_decorator.py]: I expect something like that: 



```python
spy = Mock()
function = cache(times=3)(spy)
function(1)
...
spy.assert_call...(...)
```

**Review by Cjkjvfnby — APPROVED** [review]

File: homework3/task01/project_cache/task1cache.py (hunk 6 above)
Comment (IuliiaLanchuk) [reply homework3/task01/project_cache/task1cache.py]: Yes, agree, fixed

File: homework3/task01/tests/test_cach_decorator.py (hunk 7 above)
Comment (IuliiaLanchuk) [reply homework3/task01/tests/test_cach_decorator.py]: Fixed

**Pull request merged**